add ratelimit info

oliver006 opened the pull request:
This PR makes available the rate limit information as returned by Shopify in the headers of the response to an API request. [Relevant Shopify API docs are here](https://help.shopify.com/en/api/reference/rest-admin-api-rate-limits#rate-limits-header).

I'm not sure if putting the results into a struct in the client is the best way to do this, let me know if you have suggestions for how to handle this better.

gordcurrie commented on goshopify.go:
```diff
@@ -285,6 +293,15 @@ func (c *Client) Do(req *http.Request, v interface{}) error {
+		}
+	}
+
+	c.RateLimits.RetryAfterSeconds, _ = strconv.ParseFloat(resp.Header.Get("Retry-After"), 64)
```
I'd prefer that we don't use blank identifiers for errors, we should probably handle this error.

gordcurrie replied:
After discussing with @oliver006 I see his point there's not really a good way to handle this error, returning the default `0` is probably the safest bet so there is no need to check this error.

gordcurrie commented on goshopify.go:
```diff
@@ -285,6 +293,15 @@ func (c *Client) Do(req *http.Request, v interface{}) error {
 
+	if rl := resp.Header.Get("X-Shopify-Shop-Api-Call-Limit"); rl != "" {
+		if s := strings.Split(rl, "/"); len(s) == 2 {
+			c.RateLimits.RequestCount, _ = strconv.Atoi(s[0])
```
I'd prefer that we don't use blank identifiers for errors, we should probably handle this error.

gordcurrie replied:
After discussing with @oliver006 I see his point there's not really a good way to handle this error, returning the default `0` is probably the safest bet so there is no need to check this error.

gordcurrie commented on goshopify.go:
```diff
@@ -285,6 +293,15 @@ func (c *Client) Do(req *http.Request, v interface{}) error {
+	if rl := resp.Header.Get("X-Shopify-Shop-Api-Call-Limit"); rl != "" {
+		if s := strings.Split(rl, "/"); len(s) == 2 {
+			c.RateLimits.RequestCount, _ = strconv.Atoi(s[0])
+			c.RateLimits.BucketSize, _ = strconv.Atoi(s[1])
```
I'd prefer that we don't use blank identifiers for errors, we should probably handle this error.

gordcurrie replied:
After discussing with @oliver006 I see his point there's not really a good way to handle this error, returning the default `0` is probably the safest bet so there is no need to check this error.

oliver006 commented:
Thanks for your comments.
How'd you handle the errors? As in, what would you set e.g. `RequestCount` to if the call to Atoi() fails?

gordcurrie commented:
> Thanks for your comments.
> How'd you handle the errors? As in, what would you set e.g. `RequestCount` to if the call to Atoi() fails?

I would probably return the error similar to how the error from `decoder.Decode()` is handled on line 290

oliver006 commented:
> > Thanks for your comments.
> > How'd you handle the errors? As in, what would you set e.g. `RequestCount` to if the call to Atoi() fails?
> 
> I would probably return the error similar to how the error from `decoder.Decode()` is handled on line 290

I don't think returning an error is an appropriate response. After all, only the the extraction of headers failed, the request might still have succeeded.

gordcurrie commented:
> > > r your comments.
> > > How'd you handle the errors? As in, what would you set e.g. `RequestCount` to if the call to Atoi() fails?
> > 
> > 
> > I would probably return the error similar to how the error from `decoder.Decode()` is handled on line 290
> 
> I don't think returning an error is an appropriate response. After all, only the the extraction of headers failed, the request might still have succeeded.

Valid point, and we wouldn't want people needlessly retrying writes. In that case I'm ok with returning the default values, so you're right it doesn't make sense to check those errors. I'll resolve my comments for those RequestCount and BucketSize.

What are your thoughts on the `Retry-After` header, if we can't parse it and default to 0 is that intuitive enough or will people just retry right away? I guess there aren't any other real options as we have no idea how long we should be deferring them for, the next request would likely 429 again but we have no way to know. I'll resolve my comment for that one as well, if you want to discuss further let me know.

oliver006 commented:
Thanks for the review.
Good question about the `Retry-After` header, i was going back and forth about it.
I think it's ok for it to be zero when not present.
 It's only of interest when the response code of the API request is a 429 so when absent could just default to zero. In case of a 429 response, the header should be present and it'd be up to the caller to retry after the suggested wait period. 

Rebased, resolved the merge conflict, and simplified the header value lookup - RFM

oliver006 commented:
Good call to add tests @gordcurrie - done.

gordcurrie commented:
Thanks a ton for the contribution. Merging.
